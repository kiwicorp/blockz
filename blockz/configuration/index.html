<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Support for common configuration behaviour and loading them from various sources."><meta name="keywords" content="rust, rustlang, rust-lang, configuration"><title>blockz::configuration - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../blockz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module configuration</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li></ul></div><div id="sidebar-vars" data-name="configuration" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img src="../../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">blockz</a>::<wbr><a class="mod" href="#">configuration</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/blockz/configuration.rs.html#1-274" title="goto source code">[src]</a></span></h1><div class="item-info"><div class="stab portability">This is supported on <strong>crate feature <code>configuration</code></strong> only.</div></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Support for common configuration behaviour and loading them from various sources.</p>
<h1 id="direct-configuration-example" class="section-header"><a href="#direct-configuration-example">Direct configuration example</a></h1>
<p>This example shows how to use the direct configuration. Please note that the direct
configuration is usually useful in testing scenarios. Production usage is discouraged.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">direct</span>)]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">Configuration</span><span class="op">&gt;</span><span class="ident">::load</span>(<span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="string">&quot;127.0.0.1&quot;</span>.<span class="ident">to_string</span>(),
    <span class="ident">port</span>: <span class="number">58732</span>,
}).<span class="kw">await</span>;
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to 127.0.0.1:58732.</span></pre></div>
<p>If the configuration type implements Default, you can use EasyConfiguration to load the default
value, like so:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">direct</span>)]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">impl</span> <span class="ident">Default</span> <span class="kw">for</span> <span class="ident">ServerConfig</span> {
    <span class="kw">fn</span> <span class="ident">default</span>() <span class="op">-</span><span class="op">&gt;</span> <span class="ident">ServerConfig</span> {
        <span class="ident">ServerConfig</span> {
            <span class="ident">address</span>: <span class="string">&quot;0.0.0.0&quot;</span>.<span class="ident">to_string</span>(),
            <span class="ident">port</span>: <span class="number">9999</span>,
        }
    }
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>().<span class="kw">await</span>;
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to 0.0.0.0:9999.</span></pre></div>
<h1 id="env-configuration-examples" class="section-header"><a href="#env-configuration-examples">Env configuration examples</a></h1>
<p><strong>NOTE</strong>: These examples require the <strong>env_configuration</strong> feature.</p>
<p>The first example shows how to use the env configuration with a prefix.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix</span> <span class="op">=</span> <span class="string">&quot;COOL_APP_&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
<p>The second example shows how to use the env configuration with a prefix source in the form of a
constant.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="doccomment">/// This is our prefix.</span>
<span class="kw">const</span> <span class="ident">PREFIX</span>: <span class="kw-2">&amp;</span><span class="ident">str</span> <span class="op">=</span> <span class="string">&quot;COOL_APP_&quot;</span>;

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix_source</span> <span class="op">=</span> <span class="string">&quot;PREFIX.to_string()&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
<p>The third example shows how to use the env configuration with a prefix source in the form of a
function that can return an error. The error must be mapped to an <code>envy::Error</code>.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered">
<span class="doccomment">/// This function sources our prefix.</span>
<span class="kw">fn</span> <span class="ident">source_prefix</span>() <span class="op">-</span><span class="op">&gt;</span> <span class="prelude-ty">Result</span><span class="op">&lt;</span><span class="ident">String</span>, <span class="ident">envy::Error</span><span class="op">&gt;</span> {
    <span class="ident">std::env::var</span>(<span class="string">&quot;COOL_APP_PREFIX&quot;</span>)
        .<span class="ident">map_err</span>(<span class="op">|</span><span class="ident">err</span><span class="op">|</span> <span class="ident">envy::Error::Custom</span>(<span class="macro">format!</span>(<span class="string">&quot;failed to source the prefix: {}&quot;</span>, <span class="ident">err</span>)))
}

<span class="attribute">#[<span class="ident">derive</span>(<span class="ident">Configuration</span>, <span class="ident">Deserialize</span>)]</span>
<span class="attribute">#[<span class="ident">configuration</span>(<span class="ident">env</span>(<span class="ident">prefix_source</span> <span class="op">=</span> <span class="string">&quot;source_prefix()?&quot;</span>))]</span>
<span class="kw">struct</span> <span class="ident">ServerConfig</span> {
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_addr&quot;</span>)]</span>
    <span class="ident">address</span>: <span class="ident">String</span>,
    <span class="attribute">#[<span class="ident">serde</span>(<span class="ident">rename</span> <span class="op">=</span> <span class="string">&quot;bind_port&quot;</span>)]</span>
    <span class="ident">port</span>: <span class="ident">u16</span>,
}

<span class="kw">let</span> <span class="ident">config</span> <span class="op">=</span> <span class="op">&lt;</span><span class="ident">ServerConfig</span> <span class="kw">as</span> <span class="ident">EasyConfiguration</span><span class="op">&gt;</span><span class="ident">::load</span>()
    .<span class="kw">await</span>
    .<span class="ident">expect</span>(<span class="string">&quot;Failed to load configuration from the environment!&quot;</span>);
<span class="macro">println!</span>(<span class="string">&quot;Server binding to {}:{}.&quot;</span>, <span class="ident">config</span>.<span class="ident">address</span>, <span class="ident">config</span>.<span class="ident">port</span>);
<span class="comment">// Server binding to aaa.bbb.ccc.ddd:ppppp.</span></pre></div>
</div></details><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.DirectConfiguration.html" title="blockz::configuration::DirectConfiguration struct">DirectConfiguration</a></td><td class="docblock-short"><p>Direct configuration that just returns the passed value.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.EnvConfiguration.html" title="blockz::configuration::EnvConfiguration struct">EnvConfiguration</a></td><td class="docblock-short"><span class="stab portability" title="This is supported on crate feature `env_configuration` only"><code>env_configuration</code></span><p>Configuration that can be sourced from environment variables.</p>
</td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.Configuration.html" title="blockz::configuration::Configuration trait">Configuration</a></td><td class="docblock-short"><p>Common behaviour of configurations.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.EasyConfiguration.html" title="blockz::configuration::EasyConfiguration trait">EasyConfiguration</a></td><td class="docblock-short"><p>An easy configuration is a configuration that can be loaded without other parameters.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="blockz" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div><script src="../../main.js"></script></body></html>