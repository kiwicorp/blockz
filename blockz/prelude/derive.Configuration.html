<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Derive the Configuration trait."><meta name="keywords" content="rust, rustlang, rust-lang, Configuration"><title>Configuration in blockz::prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc derive"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../blockz/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><div class="sidebar-elems"><p class="location">Other items in<br><a href="../index.html">blockz</a>::<wbr><a href="index.html">prelude</a></p><div id="sidebar-vars" data-name="Configuration" data-ty="derive" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img src="../../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Derive Macro <a href="../index.html">blockz</a>::<wbr><a href="index.html">prelude</a>::<wbr><a class="derive" href="#">Configuration</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></span></h1><pre class="rust derive">#[derive(Configuration)]
{
    // Attributes available to this derive:
    #[configuration]
}
</pre><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Derive the Configuration trait.</p>
<p>This requires that the struct or enum is <a href="https://doc.rust-lang.org/stable/std/marker/trait.Send.html">Send</a>.</p>
<p>Configuring the implementation is done via the <code>#[configuration]</code> attribute:</p>
<ul>
<li>by default, direct configuration will be implemented, if no other configuration feature is
enabled</li>
<li>direct configuration: <code>#[configuration(direct)]</code></li>
<li>env configuration with no prefix: <code>#[configuraton(env())]</code></li>
<li>env configuration with a prefix: <code>#[configuraton(env(prefix = &quot;MY_PREFIX&quot;))]</code></li>
<li>env configuration with a prefix source: <code>#[configuraton(env(prefix_source = &quot;MY_SOURCE&quot;))]</code></li>
</ul>
<p>The prefix source will be interpreted as an expression that will be used to source the prefix.
You can use either constants or functions.</p>
<p>Required available imports:</p>
<ul>
<li><a href="https://docs.rs/anyhow">anyhow</a></li>
<li><a href="https://docs.rs/async_trait">async_trait</a></li>
<li><a href="https://github.com/selftechio/blockz">blockz</a></li>
<li><a href="https://docs.rs/config">config</a></li>
</ul>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="blockz" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div><script src="../../main.js"></script></body></html>