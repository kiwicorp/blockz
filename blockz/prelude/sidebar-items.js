initSidebarItems({"attr":[["singleton_fn","Modify a method on a type that implements the Singleton trait."]],"derive":[["Configuration","Derive the Configuration trait."],["Singleton","Derive the Singleton trait."]]});