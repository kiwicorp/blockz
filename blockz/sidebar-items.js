initSidebarItems({"attr":[["singleton_fn","Modify a method on a type that implements the Singleton trait."]],"derive":[["Configuration","Derive the Configuration trait."],["Singleton","Derive the Singleton trait."]],"mod":[["configuration","Support for common configuration behaviour and loading them from various sources."],["prelude","Blockz prelude - useful re-exports."],["singleton","Support for singletons."]]});