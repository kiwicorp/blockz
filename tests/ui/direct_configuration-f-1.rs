//! Direct configuration ui test #1 - typo.

#![cfg(not(all(feature = "envy_configuration")))]

use blockz::prelude::*;

#[derive(Configuration, PartialEq)]
#[configuration(direc)]
struct EnvConfig {
    server_port: u32,
}

#[tokio::main]
async fn main() {
    panic!("This should not run!");
}
