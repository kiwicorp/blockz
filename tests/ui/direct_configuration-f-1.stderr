error: Unknown field: `direc`. Did you mean `direct`?
  --> $DIR/direct_configuration-f-1.rs:13:17
   |
13 | #[configuration(direc)]
   |                 ^^^^^
